src/traces/pie/calc.js

```diff
@@ -25,26 +24,26 @@ function calc(gd, trace) {
     var labels = trace.labels;
     var colors = trace.marker.colors || [];
     var vals = trace.values;
-    var hasVals = isArrayOrTypedArray(vals) && vals.length;
+    var len = trace._length;
+    var hasValues = trace._hasValues && len;
 
     var i, pt;
 
     if(trace.dlabel) {
-        labels = new Array(vals.length);
-        for(i = 0; i < vals.length; i++) {
+        labels = [];
```

Comment:
What's wrong with `labels = new Array(len);`. I'm not a fan an implicit array-pushing e.g.

```js
var arr = [];
for(var i = 0; i < N; i++) {
  arr[i] = i;
}

like you have here.

Reply:
Good call. Done in https://github.com/plotly/plotly.js/pull/4416/commits/1967fd459ef07a8e09edc52a997d74539bd9914a.